qdb/tracer.py

```diff
@@ -450,6 +503,109 @@ def set_quit(self):
         """
         self.quitting = True
 
+    def eval_(self, code, pprint=False):
+        repr_fn = self.repr_fn
+
+        outexc = None
+        outmsg = None
+        with capture_output() as (out, err), \
+                self._new_execution_timeout(code), \
+                self.inject_default_namespace() as stackframe:
+            try:
+                if not repr_fn and not pprint:
+                    self.eval_fn(
+                        code,
+                        stackframe,
+                        'single',
+                    )
+                else:
+                    try:
+                        # Do some some custom single mode magic that lets us
+                        # call the repr function on the last expr.
+                        value = progn(
+                            code,
+                            self.eval_fn,
+                            stackframe,
+                        )
+                    except QdbPrognEndsInStatement:
+                        # Statements have no value to print.
+                        pass
+                    else:
+                        if pprint:
+                            value = pformat(value)
+                        if repr_fn:
+                            value = repr_fn(value)
+                        print(value)
+            except Exception as e:
+                outexc = type(e).__name__
+                outmsg = self.exception_serializer(e)
+            else:
+                outmsg = (out.getvalue()[:-1] if out.getvalue() and
+                          out.getvalue()[-1] == '\n' else out.getvalue())
+
+        if outexc is not None or outmsg is not None:
+            self.cmd_manager.send_print(code, outexc, outmsg)
+
+        self.update_watchlist()
+
+    def _stack_jump_to(self, index):
+        """
+        Jumps the stack to a specific index.
+        Raises an IndexError if the desired index does not exist.
+        """
+        # Try to jump here first. This could raise an IndexError which will
+        # prevent the tracer's state from being corrupted.
+        self.curframe = self.stack[index][0]
+
+        self.curindex = index
+        self.curframe_locals = self.curframe.f_locals
+        self.update_watchlist()
+
+    def stack_shift_direction(self, direction):
+        """
+        Shifts the stack up or down depending on dirction.
+        If direction is positive, travel up, if direction is negative, travel
+        down. If direction is 0, do nothing.
+        If you cannot shift in the desired direction, an IndexError will be
+        raised.
+        """
+        if direction == 0:
+            return  # nop
+
+        direction = -1 if direction > 0 else 1
```

Comment:
I don't see how `copysign` shows the intent any better. I can understand that flipping the direction here is confusing; so maybe I should just refactor it to not need that.